sdk/metric/sdk.go

```diff
@@ -391,62 +381,91 @@ func (m *SDK) Labels(kvs ...core.KeyValue) api.LabelSet {
 		oi++
 	}
 	kvs = kvs[0:oi]
-	ls.slice = kvs
 	ls.computeOrdered(kvs)
 	return ls
 }
 
+func (ls *labels) ensureCachedValue() {
+	if !ls.cachedValue.IsValid() {
+		if ls.ordered == nil {
+			// it's an empty labelset
+			ls.cachedValue = emptyArrayValue
+		} else {
+			ls.cachedValue = reflect.ValueOf(ls.ordered)
+		}
+	}
+}
+
+func (ls *labels) NumLabels() int {
+	ls.ensureCachedValue()
+	return ls.cachedValue.Len()
+}
+
+func (ls *labels) GetLabel(idx int) core.KeyValue {
+	ls.ensureCachedValue()
+	return ls.cachedValue.Index(idx).Interface().(core.KeyValue)
```

Comment:
nit: I think this causes an allocation because a `core.KeyValue` is passing through an `interface{}`. The fix (which appears in the builder below) is to take the address, so that a pointer passes through the interface, and then dereference the pointer after it's "out" of the reflect.Value.

Reply:
Fortunately it doesn't cause an allocation. Or at least this is what benchmarks tell me. As I pointed out earlier - using Addr() in this case result in panic, as for some reason elements of the array are not addressable (see https://play.golang.org/p/IM0qJfW04HV). But I think that the non-addressability of the element allows for some optimization of not allocating a memory for element in `Interface()` call.

At least we can use the benchmarks I wrote to track this.

Reply:
Interesting. If the compiler has gotten clever, that's great. There is an explanation for the behavior in the playground. If you start with a `reflect.New()`, what you have is a pointer to an array, which makes the elements addressable. If you have an interface to an array w/o a pointer to it, it's not addressable.

Reply:
For the record, https://play.golang.org/p/_LzD-YXHlPl
placing the pointer-to-array in the interface works ^^^